### PR TITLE
Remove global state. Immutable mailer.

### PR DESCRIPTION
**Pull request by jodosha**
I [recently outlined](https://discourse.hanamirb.org/t/hanami-2-0-ideas/306/36) the principles for Hanami 2.0.

TL;DR is: proposal to remove global state cause it causes pains for testability, it limited Hanami features, and it's an obstacle for building modular applications (due to the globals). I also suggested to make all the components immutable so we can save a lot of useless objects allocations, and to make sure apps have less side effects and so less bugs.

---

⚠️ ⚠️ ⚠️ 

**I wanted to validate these ideas by converting `hanami-mailer` to follow these principles.**

**The changes of this PR are only an experiment, that's why they are targeting `unstable` branch**

⚠️ ⚠️ ⚠️ 

---

## Changes

  * All the codebase uses `# frozen_string_literal: true` pragma for Ruby 3.0
  * `Hanami::Mailer` is not a module to include, but a class to inherit from.  
  * `Hanami::Mailer.configure` and `Hanami::Mailer.configuration` are gone
  * `Hanami::Mailer::Configuration#initialize` now accepts a block
  * Settings in `Hanami::Mailer::Configuration` are not using the "DSL syntax" (eg. `root "..."`), but the "setter syntax" (`config.root = "..."`)
  * Configuration needs to be "finalized" via `configuration = Hanami::Mailer.finalize(configuration)`. This process freezes the configuration.
  * `Hanami::Mailer#initialize` now accepts an `Hanami::Mailer::Configuration` instance
  * `Hanami::Mailer#initialize` builds a frozen instance of a mailer
  * `Hanami::Mailer.deliver` is gone
  * Added `Hanami::Mailer#deliver` (aliased as `#call`). It accepts a `Hash` of locals
  * Removed "symbol DSL" for `Hanami::Mailer.subject`,`.from`, `.to`, `.cc`, `.bcc` (eg: `from :sender`)
  * Added "proc DSL" for `Hanami::Mailer.subject`,`.from`, `.to`, `.cc`, `.bcc` (eg: `from ->(locals) { locals.fetch(:user).email }`)
  * `Hanami::Mailer#prepare` is gone, in favor of `Hanami::Mailer.before`

## New Usage

```ruby
# frozen_string_literal: true
require 'bundler/setup'
require 'hanami/mailer'

configuration = Hanami::Mailer::Configuration.new do |config|
  config.root            = File.expand_path(__dir__, "base")
  config.delivery_method = :test
end

class InvoiceMailer < Hanami::Mailer
  from "invoices@domain.test"
  to ->(locals) { locals.fetch(:user).email }

  subject ->(locals) { "Invoice ##{locals.fetch(:invoice).number}" }

  before do |mail, locals|
    mail.attachments["invoice–#{locals.fetch(:invoice).number}.pdf"] = generate_invoice
  end

  private

  def generate_invoice
    # ...
  end
end

configuration = Hanami::Mailer.finalize(configuration)

invoice = Invoice.new(id: 1, number: 23)
user    = User.new(name: "Luca", email: "luca@domain.test")

mailer = InvoiceMailer.new(configuration: configuration)
mail = mailer.deliver(invoice: invoice, user: user)

mail.class # => Mail::Message
```

### Before

```
# Speed
Warming up --------------------------------------
             deliver    39.000  i/100ms
Calculating -------------------------------------
             deliver    383.163  (± 8.9%) i/s -      1.911k in   5.032289s

# Memory for 1,000 deliveries
total allocations: 2,238,013
total memsize: 188,909,210
```

### After

```
# Speed
Warming up --------------------------------------
             deliver    39.000  i/100ms
Calculating -------------------------------------
             deliver    380.200  (± 8.9%) i/s -      1.911k in   5.073797s

# Memory for 1,000 deliveries
total allocations: 2,225,013
total memsize: 186,422,863
```

Both speed and memory are the same.

---

/cc @hanami/core

### CODE REVIEW

File: lib/hanami/mailer.rb [hunk 1]
```diff
@@ -38,300 +25,265 @@ def initialize
...
-
-        self.configuration = conf.duplicate
-      end
-
-      conf.copy!(base)
+    # @since next
+    # @api unstable
+    def self.inherited(base)
+      @_subclasses.push(base)
+      base.extend Dsl
```
Comment (mereghost) [inline lib/hanami/mailer.rb]: Still trying to wrap my head around the changes, but can't we just extend `Hanami::Mailer` and not subvert the hook so it adds methods not defined on the superclass?

**Comment by sandelius** [comment]
@jodosha How would one use it in an Action?

File: lib/hanami/mailer.rb (hunk 1 above)
Comment (jodosha) [inline lib/hanami/mailer.rb]: @mereghost First of all, let me explain the idea of transforming `Hanami::Mailer` from a module to a superclass.

This new version of mailer imposes really strict rules: immutability and Single Responsibility Principle (SRP). Indeed, the only public method is `#deliver` (aliased as `#call`).

Because we want that an object that uses `Hanami::Mailer` must be only a mailer, the inheritance helps us to enforce this rule. When you have a module, you can always include it somewhere, but that breaks SRP.

---

I'm not sure what's your idea, but here's how's designed the actual code.

If you use:

```ruby
[1] pry(main)> class Mailer
[1] pry(main)*   def self.inherited(mailer)
[1] pry(main)*     mailer.class_eval do
[1] pry(main)*       @template = nil
[1] pry(main)*     end
[1] pry(main)*   end
[1] pry(main)*   def self.template(value = nil)
[1] pry(main)*     if value.nil?
[1] pry(main)*       @template
[1] pry(main)*     else
[1] pry(main)*       @template = value
[1] pry(main)*     end
[1] pry(main)*   end
[1] pry(main)* end
[2] pry(main)> class WelcomeMailer < Mailer
[2] pry(main)*   template 'welcome'
[2] pry(main)* end
"welcome"
[3] pry(main)> WelcomeMailer.template # OK
"welcome"
[4] pry(main)> Mailer.template # OK, but I don't like it
nil
```

It doesn't makes sense for `Hanami::Mailer` to expose `.template`, but only for its subclasses:

```ruby
[1] pry(main)> class Mailer
[1] pry(main)*   module Dsl
[1] pry(main)*     def self.extended(mailer)
[1] pry(main)*       mailer.class_eval do
[1] pry(main)*         @template = nil
[1] pry(main)*       end
[1] pry(main)*     end
[1] pry(main)*     def template(value = nil)
[1] pry(main)*       if value.nil?
[1] pry(main)*         @template
[1] pry(main)*       else
[1] pry(main)*         @template = value
[1] pry(main)*       end
[1] pry(main)*     end
[1] pry(main)*   end
[1] pry(main)*   def self.inherited(mailer)
[1] pry(main)*     mailer.extend(Dsl)
[1] pry(main)*   end
[1] pry(main)* end
:inherited
[2] pry(main)> class WelcomeMailer < Mailer
[2] pry(main)*   template 'welcome'
[2] pry(main)* end
"welcome"
[3] pry(main)> WelcomeMailer.template # OK
"welcome"
[4] pry(main)> Mailer.template # OK, because we don't want `Mailer.template`
NoMethodError: undefined method `template' for Mailer:Class
from (pry):24:in `__pry__'
```

Is it clear now?

**Comment by jodosha** [comment]
@sandelius For Hanami projects we're gonna experiment with an IoC container.

## Option 1

```ruby
module Web::Controllers::Users
  class Create
    include Web::Action
    import 'welcome_mailer'

    def call(params)
      # ...
      welcome_mailer.deliver(user: user)
    end
  end
end
```

## Option 2

```ruby
module Web::Controllers::Users
  class Create
    include Web::Action

    def call(params)
      # ...
      Hanami::Container['welcome_mailer'].deliver(user: user)
    end
  end
end
```

This is still open for discussion.

File: lib/hanami/mailer.rb (hunk 1 above)
Comment (mereghost) [inline lib/hanami/mailer.rb]: 👍 "I can see clearly now that the rain has gone" as Jimmy Cliff would say.

**Comment by mereghost** [comment]
To be honest I prefer the `Option 2` approach. While it seems more verbose, it's also more explicit.

**Comment by BenBach** [comment]
I also prefer option 2. I use `Dry::Container` a lot.

File: lib/hanami/mailer.rb [hunk 2]
```diff
@@ -38,300 +25,265 @@ def initialize
...
     #
-    # @see Hanami::Mailer::Configuration
+    # @see Hanami::Mailer.finalize
+    class UnknownMailerError < Error
+      # @param mailer [Hanami::Mailer] a mailer
+      #
+      # @since next
+      # @api unstable
+      def initialize(mailer)
+        super("Unknown mailer: #{mailer.inspect}. Please finalize the configuration before to use it.")
```
Comment (gizotti) [inline lib/hanami/mailer.rb]: "before using it" would be the correct grammar in this sentence 😄

File: lib/hanami/mailer.rb [hunk 3]
```diff
@@ -38,300 +25,265 @@ def initialize
...
     end
 
-    # Test deliveries
+    private_class_method :inherited
+
+    # Finalize the configuration
     #
-    # This is a collection of delivered messages, used when <tt>delivery_method</tt>
-    # is set on <tt>:test</tt>
+    # This should be used before to start to use the mailers
```
Comment (gizotti) [inline lib/hanami/mailer.rb]: "This should be used before using the mailers"  would be a better sentence than the above.

**Comment by davydovanton** [comment]
@BenBach @mereghost 
Also, we have other option how to use it in actions. It looks like the first option but without "magic" injector

### Option 3
```ruby
module Web::Controllers::Users
  class Create
    include Web::Action
    attr_reader :welcome_mailer

    def intialize(welcome_mailer: Hanami::Container['welcome_mailer'])
      @welcome_mailer = welcome_mailer
    end

    def call(params)
      # ...
      welcome_mailer.deliver(user: user)
    end
  end
end
```

In this case, we have access to welcome_mailer dependency from action constructor. And in specs, we can use what we want instead mocking container class:

```ruby
# with option 2

Hanami::Container.mock('welcome_mailer', mock_mailer)
action.call(params)

#with option 1 and 3

action = Web::Controllers::Users::Create.new(welcome_mailer: mock_mailer)
action.call(params)
```

And in this case, I prefer 1 or 3 option.
